narwhal/worker/src/handlers.rs

```diff
@@ -77,6 +78,40 @@ impl<V: TransactionValidator> WorkerToWorker for WorkerReceiverHandler<V> {
 
         Ok(anemo::Response::new(RequestBatchResponse { batch }))
     }
+
+    async fn request_batches(
+        &self,
+        request: anemo::Request<RequestBatchesRequest>,
+    ) -> Result<anemo::Response<RequestBatchesResponse>, anemo::rpc::Status> {
+        const MAX_REQUEST_BATCHES_RESPONSE_SIZE: usize = 6_000_000;
+        const MAX_REQUEST_BATCH_DIGESTS: usize = 1_000;
+
+        let mut digests_to_fetch = request.into_body().batch_digests;
+        let mut remaining_batch_digests = digests_to_fetch
```

Comment:
`RequestBatches` is only issued from primary to worker, right? So there should be no risk of malicious overloading. But we do need rate limits and flow controls. My suggestions is to keep the size limit high (6MB is ok), and have the caller / primary determine number of digests. i.e. if batches to fetch should be chunked in the primary, and we only need a sanity check here.

We can come back and update this later.

Reply:
But RequestBatches can go from primary to other workers so byzantine Primary A could overload Worker B and thus overload Primary B right?

Reply:
Got it. In this case, I think we really should return an error if the number of digests received is over a threshold.

Reply:
I can increase the limit to a number that is much larger than what we would expect and then return an error but for example even with 5k batch limit we do have nodes in normal operation with [P99 batch counts](https://mysten.grafana.net/d/ORCQSHfVk/subscriber-bulk-fetch-dashboard?var-Environment=mysten-metrics-internal&var-network=benchmark&var-validator=All&orgId=1&from=1679550342723&to=1679553942723&refresh=5s&viewPanel=5) for the `CommittedSubdag` close to 10k batches which I don't believe we should return an error for right?

Reply:
Just my arbitrary opinion, but I don't see why it's better to return an error? When there's already an interface to truncate responses based on size, why not do it based on count too? Seems simpler to keep the truncation logic all in one place rather than do it for size here and for count elsewhere.

Reply:
Alternatively, don't have a count limit at all and instead just have an internal limit within the worker on how many batches you will fetch from the DB at once? That seems like potentially best of both worlds.

Reply:
My personal preference is for each end to utilize all their information so the protocol can be as efficient as possible. It does not seem to be an efficient protocol for the requestor to send well over 1000 batch digests in the request, and have everything after index 1000 sent back in the response. There are two areas I discussed with @arun-koshy 
1. It should be straightforward for the requestor to limit the number of batch digests in the request. In future we can also have parallel fetching but that is not needed for now.
2. The response contains a list of digests that exist on the worker but not included because of exceeding size limit. This is useful when fetching from own worker, but cannot be relied upon when fetching from other workers. We need to be careful when using this information.

Reply:
> My personal preference is for each end to utilize all their information so the protocol can be as efficient as possible. It does not seem to be an efficient protocol for the requestor to send well over 1000 batch digests in the request, and have everything after index 1000 sent back in the response. There are two areas I discussed with @arun-koshy
> 
> 1. It should be straightforward for the requestor to limit the number of batch digests in the request. In future we can also have parallel fetching but that is not needed for now.
> 2. The response contains a list of digests that exist on the worker but not included because of exceeding size limit. This is useful when fetching from own worker, but cannot be relied upon when fetching from other workers. We need to be careful when using this information.

Fair points, Mingwei. I would recommend the following changes.
1. Entirely remove the batch count limit in RPC handler.
2. In its place, add an internal batch fetch limit. e.g. Fetch at most, say, ~200 batches at a time from DB. Add each fetched group to the response and terminate early if response size limit is exceeded. This limits RAM use exposure for malicious requester, while also allowing unlimited # of batches to be in one response in the more common case where they're all small enough to fit.
3. Remove the missing digest from response, the caller has to recalculate digests of what it got back anyway so it can decide on its own what was missing.

I think this addresses all three concerns?
- protocol efficiency
- batching as much as possible in each response, up to the limit
- protection against malicious peers